analyzers/src/SonarAnalyzer.CSharp/Rules/SymbolicExecution/SymbolicExecutionRunner.cs

```diff
@@ -98,62 +111,83 @@ protected override void Initialize(SonarAnalysisContext context)
             }
         }
 
-        private void Analyze(SonarAnalysisContext analysisContext, SyntaxNodeAnalysisContext context, CSharpSyntaxNode body, ISymbol symbol)
+        private void Analyze(SonarAnalysisContext sonarContext, SyntaxNodeAnalysisContext nodeContext, CSharpSyntaxNode body, ISymbol symbol)
         {
-            var isTestProject = analysisContext.IsTestProject(context.Compilation, context.Options);
-            var isScannerRun = analysisContext.IsScannerRun(context.Options);
-            var enabledAnalyzers = symbolicExecutionAnalyzerFactory.GetEnabledAnalyzers(context, isTestProject, isScannerRun);
-            if (body == null
-                || body.ContainsDiagnostics
-                || !enabledAnalyzers.Any()
-                || !CSharpControlFlowGraph.TryGet(body, context.SemanticModel, out var cfg))
+            if (body != null && !body.ContainsDiagnostics)
             {
-                return;
+                var isTestProject = sonarContext.IsTestProject(nodeContext.Compilation, nodeContext.Options);
+                var isScannerRun = sonarContext.IsScannerRun(nodeContext.Options);
+                AnalyzeSonar(nodeContext, isTestProject, isScannerRun, body, symbol);
+                if (ControlFlowGraph.IsAvailable)   // ToDo: Make this configurable for UTs when migrating other rules, see DeadStores
+                {
+                    AnalyzeRoslyn(sonarContext, nodeContext, isTestProject, isScannerRun, body, symbol);
+                }
             }
+        }
 
-            var lva = new SonarCSharpLiveVariableAnalysis(cfg, symbol, context.SemanticModel);
-            try
+        private void AnalyzeRoslyn(SonarAnalysisContext sonarContext, SyntaxNodeAnalysisContext nodeContext, bool isTestProject, bool isScannerRun, CSharpSyntaxNode body, ISymbol symbol)
+        {
+            var checks = AllRules.Concat(additionalTestRules)
+                .Where(x => SymbolicExecutionAnalyzerFactory.IsEnabled(nodeContext, isTestProject, isScannerRun, x.Key))
+                .GroupBy(x => x.Value.Type)                             // Multiple DiagnosticDescriptors (S2583, S2589) can share the same check type
+                .Select(x => x.First().Value.CreateInstance(sonarContext, nodeContext))   // We need just one instance in that case
+                .Where(x => x.ShouldExecute())
+                .ToArray();
+            if (checks.Any())
             {
-                var explodedGraph = new SonarExplodedGraph(cfg, symbol, context.SemanticModel, lva);
-                var analyzerContexts = enabledAnalyzers.Select(x => x.CreateContext(explodedGraph, context)).ToList();
                 try
                 {
-                    explodedGraph.ExplorationEnded += ExplorationEndedHandler;
-                    explodedGraph.Walk();
+                    var cfg = body.CreateCfg(nodeContext.SemanticModel);
+                    var engine = new RoslynSymbolicExecution(cfg, checks);
```

Comment:
From a performance view it is understandable that we run all the checks from all the rules in a single instance of the Engine. On the other hand if we mess up something in one of the rules it will affect all the others. We might revisit this decision in the future if we see that  some rules have negative impact on others.

Reply:
Previously, each rule run its own engine instance and there was a major performance penalty for that. By running all rules on a single pass was a major improvement.